x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/TransportEsqlQueryAction.java

```diff
@@ -115,22 +115,26 @@ public TransportEsqlQueryAction(
     protected void doExecute(Task task, EsqlQueryRequest request, ActionListener<EsqlQueryResponse> listener) {
         listener = listener.delegateFailureAndWrap(ActionListener::respondAndRelease);
         if (requestIsAsync(request)) {
-            asyncTaskManagementService.asyncExecute(
-                request,
-                request.waitForCompletionTimeout(),
-                request.keepAlive(),
-                request.keepOnCompletion(),
-                listener
-            );
+            requestExecutor.execute(ActionRunnable.wrap(listener, l -> asyncExecute(request, l)));
         } else {
             // workaround for https://github.com/elastic/elasticsearch/issues/97916 - TODO remove this when we can
             requestExecutor.execute(ActionRunnable.wrap(listener, l -> doExecuteForked(task, request, l)));
         }
     }
 
+    void asyncExecute(EsqlQueryRequest request, ActionListener<EsqlQueryResponse> listener) {
+        asyncTaskManagementService.asyncExecute(
+            request,
+            request.waitForCompletionTimeout(),
+            request.keepAlive(),
+            request.keepOnCompletion(),
+            listener
+        );
+    }
+
     @Override
     public void execute(EsqlQueryRequest request, EsqlQueryTask task, ActionListener<EsqlQueryResponse> listener) {
-        doExecuteForked(task, request, listener);
+        ActionListener.run(listener, l -> doExecuteForked(task, request, l));
```

Comment:
Hmm this looks like a bug to me, introduced in #103398: it's running `doExecuteForked` on the transport thread (i.e. without actually forking). Transport threads are extremely latency-sensitive, and this stuff all looks rather complicated so even if it's ok today there's a massive risk that we introduce some cluster-destroying latency in a future change. Please could we stay off the transport threads here. In particular I think we should add an `assert ThreadPool.assertCurrentThreadPool(EsqlPlugin.ESQL_THREAD_POOL_NAME);` to the top of `doExecuteForked` to make sure it's on the right thread, and then revert `doExecute` back to a method which just forks and calls `doExecuteForked`, with the logic on which you're working all happening after forking?

Reply:
I agree with keeping execution off the transport threads. I added an assertion as suggested to ensure this - it would have caught the bug when adding the async support. It's there now 👍 

I'm not sure that I follow your other suggestions. The `execute` method is required as part of the `AsyncOperation`. All it does here is to ensure that the operation catches and handles all Exceptions (this is the fix for parse/verification errors) while delegating to doExecuteForked. Async needs an indirection through the `AsyncTaskManagementService`.

Reply:
Ah ok I see the confusion, I thought this was overriding `org.elasticsearch.action.support.TransportAction#execute` but in fact it's overriding `org.elasticsearch.xpack.ql.async.AsyncTaskManagementService.AsyncOperation#execute`, the only difference between the two methods being the order of the arguments. That's pretty trappy given how differently they're called!

I'd still rather the body of `doExecute` was just `requestExecutor.execute(ActionRunnable.wrap(listener.delegateFailureAndWrap(ActionListener::respondAndRelease), l -> doExecuteForked(task, request, l)));` similar to the other places we workaround #97916. Something like this perhaps (as a patch on ada745b5970863e80bc5a1aa003db56ff7f558c7):

```diff
diff --git a/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/TransportEsqlQueryAction.java b/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/TransportEsqlQueryAction.java
index de996771d70..54edf1370e0 100644
--- a/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/TransportEsqlQueryAction.java
+++ b/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/TransportEsqlQueryAction.java
@@ -113,12 +113,20 @@ public class TransportEsqlQueryAction extends HandledTransportAction<EsqlQueryRe

     @Override
     protected void doExecute(Task task, EsqlQueryRequest request, ActionListener<EsqlQueryResponse> listener) {
-        listener = listener.delegateFailureAndWrap(ActionListener::respondAndRelease);
+        // workaround for https://github.com/elastic/elasticsearch/issues/97916 - TODO remove this when we can
+        requestExecutor.execute(
+            ActionRunnable.wrap(
+                listener.<EsqlQueryResponse>delegateFailureAndWrap(ActionListener::respondAndRelease),
+                l -> doExecuteForked(task, request, l)
+            )
+        );
+    }
+
+    private void doExecuteForked(Task task, EsqlQueryRequest request, ActionListener<EsqlQueryResponse> listener) {
         if (requestIsAsync(request)) {
-            requestExecutor.execute(ActionRunnable.wrap(listener, l -> asyncExecute(request, l)));
+            asyncExecute(request, listener);
         } else {
-            // workaround for https://github.com/elastic/elasticsearch/issues/97916 - TODO remove this when we can
-            requestExecutor.execute(ActionRunnable.wrap(listener, l -> doExecuteForked(task, request, l)));
+            innerExecute(task, request, listener);
         }
     }

@@ -134,10 +142,10 @@ public class TransportEsqlQueryAction extends HandledTransportAction<EsqlQueryRe

     @Override
     public void execute(EsqlQueryRequest request, EsqlQueryTask task, ActionListener<EsqlQueryResponse> listener) {
-        ActionListener.run(listener, l -> doExecuteForked(task, request, l));
+        ActionListener.run(listener, l -> innerExecute(task, request, l));
     }

-    private void doExecuteForked(Task task, EsqlQueryRequest request, ActionListener<EsqlQueryResponse> listener) {
+    private void innerExecute(Task task, EsqlQueryRequest request, ActionListener<EsqlQueryResponse> listener) {
         assert ThreadPool.assertCurrentThreadPool(EsqlPlugin.ESQL_THREAD_POOL_NAME);
         EsqlConfiguration configuration = new EsqlConfiguration(
             ZoneOffset.UTC,
```

Reply:
That is indeed cleaner. Done [ef89158](https://github.com/elastic/elasticsearch/pull/103755/commits/ef891589609a990ce8359e98407f6714f249ff97)